lib/semantics/resolve-names.cc

```diff
@@ -420,18 +425,127 @@ class SubprogramVisitor : public InterfaceVisitor {
   // Function result name from parser::Suffix, if any.
   const parser::Name *funcResultName_{nullptr};
 
-  bool BeginSubprogram(const parser::Name &,
+  bool BeginSubprogram(const parser::Name &, Symbol::Flag,
       const std::optional<parser::InternalSubprogramPart> &);
   void EndSubprogram();
   // Create a subprogram symbol in the current scope and push a new scope.
-  Symbol &PushSubprogramScope(const parser::Name &);
+  Symbol &PushSubprogramScope(const parser::Name &, Symbol::Flag);
   Symbol *GetSpecificFromGeneric(const parser::Name &);
 };
 
+class DeclarationVisitor : public ArraySpecVisitor,
```

Comment:
OO design point: is there an "is a" relationship between DeclarationVisitor and its base classes, or are they being used to mix in some capabilities?  Encapsulation might be clearer in the latter case.